File: WhatsDAM_Java/client/src/main/java/com/ieseljust/psp/client/communications/communicationManager.java
package com.ieseljust.psp.client.communications;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.InetSocketAddress;
import java.net.Socket;

import com.ieseljust.psp.client.CurrentConfig;
import com.ieseljust.psp.client.Message;

import org.json.JSONException;
import org.json.JSONObject;

public class communicationManager {

    public static JSONObject sendServer(String msg) {
        try {
            Socket socket = new Socket();
            socket.connect(new InetSocketAddress(CurrentConfig.getServerAddress(), CurrentConfig.getServerPort()));

            // Envia el missatge al servidor
            OutputStream outputStream = socket.getOutputStream();
            PrintWriter out = new PrintWriter(new OutputStreamWriter(outputStream));
            out.println(msg);
            out.flush();

            // Rep la resposta del servidor
            InputStream inputStream = socket.getInputStream();
            BufferedReader in = new BufferedReader(new InputStreamReader(inputStream));
            String response = in.readLine();

            // Tanca la connexió
            socket.close();

            // Parsa la resposta JSON
            return new JSONObject(response);
        } catch (Exception e) {
            e.printStackTrace();
            // Maneja errors de connexió o JSON
            return null;
        }
    }

    public static void connect() throws JSONException, communicationManagerException {
        try {
            JSONObject registerData = new JSONObject();
            registerData.put("command", "register");
            registerData.put("user", CurrentConfig.getUsername());
            registerData.put("listenPort", CurrentConfig.getListenPort());

            // Envia el missatge de registre al servidor
            sendServer(registerData.toString());
        } catch (Exception e) {
            e.printStackTrace();
            // Maneja errors de connexió o JSON
            throw new communicationManagerException("Error durant el registre");
        }
    }

    public static void sendMessage(Message m) {
        // Implementa l'enviament d'un missatge al servidor
    }

    public static class communicationManagerException extends Exception {
        public communicationManagerException(String message) {
            super(message);
        }
    }
}

File: WhatsDAMLolliPop/app/src/main/java/com/example/whatsdamlolipop/model/CommunicationManager.kt
package com.example.whatsdamlolipop.model

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.ServerSocket
import java.net.Socket

object CommunicationManager {

    private const val PORT = 9999
    private var server: String? = null
    private var listenPort: Int? = null

    suspend fun enviarServidor(msg: String): JSONObject {
        return withContext(Dispatchers.IO) {
            val socket = Socket(server, PORT)
            val writer = OutputStreamWriter(socket.getOutputStream())
            val reader = BufferedReader(InputStreamReader(socket.getInputStream()))

            try {
                // Enviar mensaje al servidor
                writer.write(msg)
                writer.flush()

                // Leer la respuesta del servidor
                val response = StringBuilder()
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    response.append(line)
                }

                // Devolver la respuesta en formato JSON
                JSONObject(response.toString())
            } finally {
                // Cerrar streams y socket
                writer.close()
                reader.close()
                socket.close()
            }
        }
    }

    suspend fun prepareListener() {
        listenPort = getFreePort()
        GlobalScope.launch(Dispatchers.IO) {
            val serverSocket = ServerSocket(listenPort!!)
            while (true) {
                val clientSocket = serverSocket.accept()
                GlobalScope.launch(Dispatchers.IO) {
                    processNotification(clientSocket)
                }
            }
        }
    }

    private suspend fun processNotification(socket: Socket) {
        val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
        val message = reader.readLine()

        // Procesar la notificación
        val jsonMessage = JSONObject(message)
        val messageType = jsonMessage.getString("type")

        when (messageType) {
            "message" -> {
                // Procesar el mensaje y agregarlo a la lista de mensajes
                val content = jsonMessage.getString("content")
                Missatges.add("Remitente", content)

                // Enviar respuesta al servidor
                val response = JSONObject().put("status", "ok")
                val writer = OutputStreamWriter(socket.getOutputStream())
                writer.write(response.toString())
                writer.flush()

                // Cerrar streams y socket
                writer.close()
                reader.close()
                socket.close()
            }
        }
    }

    suspend fun login(usuario: String, contrasena: String): JSONObject {
        return try {
            prepareListener()
            val registerMessage = JSONObject().apply {
                put("command", "register")
                put("user", usuario)
                put("listenPort", listenPort)
            }
            enviarServidor(registerMessage.toString())
        } catch (e: Exception) {
            JSONObject().put("status", "error")
        }
    }

    private fun getFreePort(): Int {
        return try {
            val serverSocket = ServerSocket(0)
            val port = serverSocket.localPort
            serverSocket.close()
            port
        } catch (e: Exception) {
            throw RuntimeException("Error al obtener un puerto disponible", e)
        }
    }
}